src/cloudevents/core/v1/exceptions.py

```diff
@@ -0,0 +1,27 @@
+#  Copyright 2018-Present The CloudEvents Authors
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+class CloudEventValidationError(Exception):
+    """
+    Custom exception for validation errors.
+    """
+
+    def __init__(self, errors: dict[str, list[str]]) -> None:
+        super().__init__("Validation errors occurred")
+        self.errors: dict[str, list[str]] = errors
+
+    def __str__(self) -> str:
+        error_messages = [
+            f"{key}: {', '.join(value)}" for key, value in self.errors.items()
+        ]
+        return f"{super().__str__()}: {', '.join(error_messages)}"
```

Comment:
```suggestion
class CloudEventValidationError(Exception):
    """
    Holds validation errors aggregated during a CloudEvent creation.
    """

    def __init__(self, errors: dict[str, list[str]]) -> None:
        super().__init__("Validation errors occurred")
        self.errors: dict[str, list[str]] = errors
        """
        The errors gathered during the CloudEvent creation where key is the name of the attribute and
        value is a list of errors related to that attribute.
        """

    def __str__(self) -> str:
        error_messages = [
            f"{key}: {', '.join(value)}" for key, value in self.errors.items()
        ]
        return f"{super().__str__()}: {', '.join(error_messages)}"
```

Reply:
I've re-read the validation part. I suggest we change the validation to actually have `errors` where dict key is the attribute name and value holds all the errors related to that attribute.

Reply:
I've been thinking about this, but do you think this will be valuable, besides grouping the errors more logically, by the field 🤔 
I'm mostly thinking about programmatically reacting to these errors when using the library. Say I get a CloudEvent via HTTP, and I'm trying to use this library and I have some errors. I can handle them in code by checking the `required` or `type` or even `optional` fields and decide programmatically how to deal with that bad event, rather than having to check errors by substrings and based on fields. 
I'm 50/50 between these 2 approaches. Let me know.

Reply:
If we're talking about some automated actions, then we can use actual typed errors instead of strings. That'd give the context of whether it's a type error or smth else + we keep the by attribute name grouping

Reply:
Let me see then, I think introducing the typed errors might be beneficial. Something like:
```
{
    "id": [TypeError("'id' field should be string"), MissingRequiredAttribute("'id'" attribute is required")]
    ...
```